File: java/app/src/main/java/net/fbvictorhugo/j/barreirasanitaria/data/dao/dummy/DummyQuestionarioDAO.java
package net.fbvictorhugo.j.barreirasanitaria.data.dao.dummy;

import net.fbvictorhugo.j.barreirasanitaria.data.dao.QuestionarioDAO;
import net.fbvictorhugo.j.barreirasanitaria.data.model.Questionario;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

public class DummyQuestionarioDAO implements QuestionarioDAO {

    private static DummyQuestionarioDAO sInstance;
    private List<Questionario> mDummyList;

    public static DummyQuestionarioDAO getInstance() {
        if (sInstance == null) {
            sInstance = new DummyQuestionarioDAO();
        }
        return sInstance;
    }

    private DummyQuestionarioDAO() {
        mDummyList = new ArrayList<>();
        buildListaFake();
    }

    private long getProximoIdDisponivel() {
        if (mDummyList.size() == 0) {
            return 1;
        } else {
            return mDummyList.get(mDummyList.size() - 1).getId() + 1;
        }
    }

    @Override
    public List<Questionario> listar() {
        return mDummyList;
    }

    @Override
    public void inserir(Questionario questionario) throws Exception {
        if (questionario == null) {
            throw new NullPointerException("Modelo Questionario está nulo.");
        }

        questionario.setId(getProximoIdDisponivel());
        mDummyList.add(questionario);
    }

    @Override
    public Questionario procurarUltimoQuestionarioRespondidoPorPessoa(long idPessoa) {
        Date lasDate = null;
        Questionario questionario = null;
        for (Questionario itemQuestionario : mDummyList) {
            if (itemQuestionario.getPessoaId() == idPessoa) {
                if (lasDate == null) {
                    lasDate = itemQuestionario.getDataResposta();
                }

                if (itemQuestionario.getDataResposta().compareTo(lasDate) >= 0) {
                    lasDate = itemQuestionario.getDataResposta();
                    questionario = itemQuestionario;
                }
            }
        }
        return questionario;
    }

    void buildListaFake() {
        //TODO LISTA FAKE
        Questionario q;

        q = new Questionario();
        q.setId(3);
        q.setPessoaId(2);
        q.setBarreiraId(1);
        q.setViagemExterior(randomBoolean());
        q.setSintomaFebre(randomBoolean());
        q.setSintomaCoriza(randomBoolean());
        q.setSintomaTosse(randomBoolean());
        q.setSintomaCancaco(randomBoolean());
        q.setSintomaDorGarganta(randomBoolean());
        q.setSintomaFaltaAr(randomBoolean());
        q.setSintomaContatoComEnfermos(randomBoolean());
        q.setDataResposta(new Date());
        mDummyList.add(q);

        q = new Questionario();
        q.setId(2);
        q.setPessoaId(2);
        q.setBarreiraId(2);
        q.setViagemExterior(randomBoolean());
        q.setSintomaFebre(randomBoolean());
        q.setSintomaCoriza(randomBoolean());
        q.setSintomaTosse(randomBoolean());
        q.setSintomaCancaco(randomBoolean());
        q.setSintomaDorGarganta(randomBoolean());
        q.setSintomaFaltaAr(randomBoolean());
        q.setSintomaContatoComEnfermos(randomBoolean());
        q.setDataResposta(new Date());
        mDummyList.add(q);

        q = new Questionario();
        q.setId(1);
        q.setPessoaId(1);
        q.setBarreiraId(1);
        q.setViagemExterior(randomBoolean());
        q.setSintomaFebre(randomBoolean());
        q.setSintomaCoriza(randomBoolean());
        q.setSintomaTosse(randomBoolean());
        q.setSintomaCancaco(randomBoolean());
        q.setSintomaDorGarganta(randomBoolean());
        q.setSintomaFaltaAr(randomBoolean());
        q.setSintomaContatoComEnfermos(randomBoolean());
        q.setDataResposta(new Date());
        mDummyList.add(q);

        q = new Questionario();
        q.setId(4);
        q.setPessoaId(1);
        q.setBarreiraId(1);
        q.setViagemExterior(randomBoolean());
        q.setSintomaFebre(randomBoolean());
        q.setSintomaCoriza(randomBoolean());
        q.setSintomaTosse(randomBoolean());
        q.setSintomaCancaco(randomBoolean());
        q.setSintomaDorGarganta(randomBoolean());
        q.setSintomaFaltaAr(randomBoolean());
        q.setSintomaContatoComEnfermos(randomBoolean());
        q.setDataResposta(new Date());
        mDummyList.add(q);

    }

    private boolean randomBoolean() {
        return Math.random() < 0.5;
    }

}

File: kotlin/app/src/main/java/net/fbvictorhugo/k/barreirasanitaria/data/dao/dummy/DummyQuestionarioDAO.kt
package net.fbvictorhugo.k.barreirasanitaria.data.dao.dummy

import net.fbvictorhugo.k.barreirasanitaria.data.dao.QuestionarioDAO
import net.fbvictorhugo.k.barreirasanitaria.data.model.Questionario
import java.util.*
import kotlin.collections.ArrayList

object DummyQuestionarioDAO : QuestionarioDAO {

    private val _dummyList: ArrayList<Questionario> = ArrayList()

    fun getProximoIdDisponivel(): Long {
        return if (_dummyList.size == 0) {
            1
        } else {
            _dummyList[_dummyList.size - 1].id + 1
        }
    }

    override fun listar(): ArrayList<Questionario> {
        return _dummyList
    }

    override fun inserir(questionario: Questionario) {
        questionario.id = getProximoIdDisponivel()
        _dummyList.add(questionario)
    }

    override fun procurarUltimoQuestionarioRespondidoPorPessoa(idPessoa: Long): Questionario? {
        return _dummyList
            .sortedByDescending { it.dataResposta }
            .find {
                it.pessoaId == idPessoa
            }
    }

    init {
        val quest = Questionario(1, 1, 1).apply {
            dataResposta = Date()
            viagemExterior = randomBoolean()
            sintomaFebre = randomBoolean()
            sintomaCoriza = randomBoolean()
            sintomaTosse = randomBoolean()
            sintomaCancaco = randomBoolean()
            sintomaDorGarganta = randomBoolean()
            sintomaFaltaAr = randomBoolean()
            sintomaContatoComEnfermos = randomBoolean()
        }
        val quest2 = Questionario(2, 2, 2).apply {
            dataResposta = Date()
            viagemExterior = randomBoolean()
            sintomaFebre = randomBoolean()
            sintomaCoriza = randomBoolean()
            sintomaTosse = randomBoolean()
            sintomaCancaco = randomBoolean()
            sintomaDorGarganta = randomBoolean()
            sintomaFaltaAr = randomBoolean()
            sintomaContatoComEnfermos = randomBoolean()
        }
        val quest3 = Questionario(3, 1, 2).apply {
            dataResposta = Date()
            viagemExterior = randomBoolean()
            sintomaFebre = randomBoolean()
            sintomaCoriza = randomBoolean()
            sintomaTosse = randomBoolean()
            sintomaCancaco = randomBoolean()
            sintomaDorGarganta = randomBoolean()
            sintomaFaltaAr = randomBoolean()
            sintomaContatoComEnfermos = randomBoolean()
        }
        val quest4 = Questionario(4, 1, 1).apply {
            dataResposta = Date()
            viagemExterior = randomBoolean()
            sintomaFebre = randomBoolean()
            sintomaCoriza = randomBoolean()
            sintomaTosse = randomBoolean()
            sintomaCancaco = randomBoolean()
            sintomaDorGarganta = randomBoolean()
            sintomaFaltaAr = randomBoolean()
            sintomaContatoComEnfermos = randomBoolean()
        }

        _dummyList.add(quest4)
        _dummyList.add(quest3)
        _dummyList.add(quest2)
        _dummyList.add(quest)

    }

    private fun randomBoolean(): Boolean {
        return Math.random() < 0.5
    }

}
